File: backend/conversation-service/src/main/java/com/fellowgrad/conversationservice/controller/ConversationController.java
package com.fellowgrad.conversationservice.controller;

import com.fellowgrad.conversationservice.dto.*;
import com.fellowgrad.conversationservice.service.ConversationService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/conversations")
@RequiredArgsConstructor
public class ConversationController {

    private final ConversationService service;

    @PostMapping
    public ResponseEntity<ConversationResponse> createConversation(
            @Valid @RequestBody CreateConversationRequest request
    ) {
        return ResponseEntity.ok(service.createConversation(request));
    }

    @GetMapping("/user/{userId}")
    public ResponseEntity<List<ConversationResponse>> getConversationsByUser(@PathVariable String userId) {
        return ResponseEntity.ok(service.getConversationsByUser(userId));
    }

    @GetMapping("/{conversationId}")
    public ResponseEntity<ConversationResponse> getConversation(@PathVariable String conversationId) {
        return ResponseEntity.ok(service.getConversation(conversationId));
    }

    @PostMapping("/{conversationId}/messages")
    public ResponseEntity<MessageResponse> sendMessage(
            @PathVariable String conversationId,
            @Valid @RequestBody SendMessageRequest request
    ) {
        return ResponseEntity.ok(service.saveMessage(conversationId, request));
    }

    @GetMapping("/{conversationId}/messages")
    public ResponseEntity<List<MessageResponse>> getMessages(@PathVariable String conversationId) {
        return ResponseEntity.ok(service.getMessages(conversationId));
    }

    @GetMapping("/{conversationId}/recent")
    public ResponseEntity<List<MessageResponse>> getRecentMessages(
            @PathVariable String conversationId,
            @RequestParam(defaultValue = "10") int limit
    ) {
        return ResponseEntity.ok(service.getRecentMessages(conversationId, limit));
    }
}

File: backend/src/controllers/conversationController.js
const ConversationService = require('../services/conversationService');

const createConversation = async (req, res, next) => {
  try {
    const { userId, title } = req.body;
    const conversation = await ConversationService.createConversation({ userId, title });
    return res.status(200).json(conversation);
  } catch (err) {
    next(err);
  }
};

const getConversationsByUser = async (req, res, next) => {
  try {
    const { userId } = req.params;
    const conversations = await ConversationService.getConversationsByUser(userId);
    return res.status(200).json(conversations);
  } catch (err) {
    next(err);
  }
};

const getConversation = async (req, res, next) => {
  try {
    const { conversationId } = req.params;
    const conversation = await ConversationService.getConversation(conversationId);
    return res.status(200).json(conversation);
  } catch (err) {
    next(err);
  }
};

const sendMessage = async (req, res, next) => {
  try {
    const { conversationId } = req.params;
    const { role, content, messageType } = req.body;
    const message = await ConversationService.saveMessage(conversationId, {
      role,
      content,
      messageType,
    });
    return res.status(200).json(message);
  } catch (err) {
    next(err);
  }
};

const getMessages = async (req, res, next) => {
  try {
    const { conversationId } = req.params;
    const messages = await ConversationService.getMessages(conversationId);
    return res.status(200).json(messages);
  } catch (err) {
    next(err);
  }
};

const getRecentMessages = async (req, res, next) => {
  try {
    const { conversationId } = req.params;
    const limit = req.query.limit || 10;
    const messages = await ConversationService.getRecentMessages(conversationId, limit);
    return res.status(200).json(messages);
  } catch (err) {
    next(err);
  }
};

module.exports = {
  createConversation,
  getConversationsByUser,
  getConversation,
  sendMessage,
  getMessages,
  getRecentMessages,
};
